jmolecules-events/src/main/java/org/jmolecules/event/annotation/DomainEventPublisher.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jmolecules.event.annotation;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Identifies a domain event publisher, i.e. logic to publish a {@link DomainEvent}.
+ *
+ * @author Christian Stettler
+ * @author Henning Schwentner
+ * @author Stephan Pirnbaum
+ * @author Martin Schimak
+ * @author Oliver Drotbohm
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
```

Comment:
I'd also like to understand better what the usage model looks like here? Annotating a class feels a bit weird as that's a pretty broad brush. Also, it would somehow need to make the user declare stuff three times: first by writing the code that actually publishes the event, which – second – is likely to be annotated with `@DomainEvent` or implementing `DomainEvent` already and then additionally declare `@DomainEventPublisher` on the class containing that logic.

I don't even think that domain event publisher is a role that can be assigned to a class. It is more the role of a method as fulfilling that role requires some interaction like calling some infrastructure, reporting the events etc. Even for that, I am not sure how useful the annotation is in the first place as the role itself immediately begs the question of "What event(s) is/are published?" which would need some deeper inspection of the actual code *or* the repetitive declaration of that information inside the annotation.

Reply:
I agree with the aspect on the method-level annotation.

Regarding the identification of the publisher: It'd be optional to use, but it clearly identifies the role of a specific method. With that, one would be able to easily identify places where commands/events are published. This is, unfortunately, not always clear otherwise.

Reply:
That's a bit of an issue I see in all of the `…Publisher` annotations: the achieved increase in abstraction is rather low and it feels like we're nudging users into stating their intention twice. Assuming there was a rule implied like "domain events / event publishing infrastructure must only be used in methods that are annotated with `@DomainEventPublisher`". That's kind of urging the user to write something like this:

```java
@DomainEventpublisher
void someMethod() {
  // yada yada
  doSomething(new MyDomainEvent(…));
}
```
I am afraid that users would at some point just play the system by slapping that annotation on just every method that instantiates event types or calls publishing infrastructure.

If we make the annotation optional, wouldn't you then have to implement the scan for event creation anyway to make sure you detect the publishers reliably anyway?

Reply:
see: https://github.com/xmolecules/jmolecules/pull/39#discussion_r545940451